s3_scrubber/src/checks.rs

```diff
@@ -160,52 +159,83 @@ impl BranchCheckStats {
     }
 }
 
-async fn branch_cleanup_and_check_errors(
-    id: TenantTimelineId,
+pub struct TimelineAnalysis {
+    /// Anomalies detected
+    pub errors: Vec<String>,
+
+    /// Healthy-but-noteworthy, like old-versioned structures that are readable but
+    /// worth reporting for awareness that we must not remove that old version decoding
+    /// yet.
+    pub warnings: Vec<String>,
+
+    /// Keys not referenced in metadata: candidates for removal
+    pub garbage_keys: Vec<String>,
+}
+
+impl TimelineAnalysis {
+    fn new() -> Self {
+        Self {
+            errors: Vec::new(),
+            warnings: Vec::new(),
+            garbage_keys: Vec::new(),
+        }
+    }
+}
+
+pub async fn branch_cleanup_and_check_errors(
+    id: &TenantTimelineId,
     s3_root: &RootTarget,
-    s3_active_branch: &BranchData,
+    s3_active_branch: Option<&BranchData>,
     console_branch: Option<BranchData>,
     s3_data: Option<S3TimelineBlobData>,
-) -> Vec<String> {
-    info!(
-        "Checking timeline for branch branch {:?}/{:?}",
-        s3_active_branch.project_id, s3_active_branch.id
-    );
-    let mut branch_check_errors = Vec::new();
-
-    match console_branch {
-        Some(console_active_branch) => {
-            if console_active_branch.deleted {
-                branch_check_errors.push(format!("Timeline has deleted branch data in the console (id = {:?}, project_id = {:?}), recheck whether if it got removed during the check",
-                    s3_active_branch.id, s3_active_branch.project_id))
-            }
-        },
-        None => branch_check_errors.push(format!("Timeline has no branch data in the console (id = {:?}, project_id = {:?}), recheck whether if it got removed during the check",
+) -> TimelineAnalysis {
+    let mut result = TimelineAnalysis::new();
+
+    info!("Checking timeline {id}");
+
+    if let Some(s3_active_branch) = s3_active_branch {
+        info!(
+            "Checking console status for timeline for branch branch {:?}/{:?}",
+            s3_active_branch.project_id, s3_active_branch.id
+        );
+        match console_branch {
+            Some(_) => {result.errors.push(format!("Timeline has deleted branch data in the console (id = {:?}, project_id = {:?}), recheck whether if it got removed during the check",
```

Comment:
You are the native speaker, but I think in "whether if" one of the two words needs to go.

Reply:
This is existing code that got re-indented, but I'll clean it up while we're here.